validator/db/proposal_history_test.go

```diff
@@ -0,0 +1,221 @@
+package db
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/prysmaticlabs/go-bitfield"
+	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+func TestProposalHistory_InitializesNewPubKeys(t *testing.T) {
+	pubkeys := [][48]byte{[48]byte{30}, [48]byte{25}, [48]byte{20}}
+	db := SetupDB(t, pubkeys)
+	defer TeardownDB(t, db)
+
+	for _, pub := range pubkeys {
+		proposalHistory, err := db.ProposalHistory(pub[:])
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		clean := &slashpb.ProposalHistory{
+			EpochBits: bitfield.NewBitlist(params.BeaconConfig().WeakSubjectivityPeriod),
+		}
+		if !reflect.DeepEqual(proposalHistory, clean) {
+			t.Fatalf("Expected proposal history epoch bits to be empty, received %v", proposalHistory)
+		}
+	}
+}
+
+func TestProposalHistory_NilDB(t *testing.T) {
+	db := SetupDB(t, [][48]byte{})
+	defer TeardownDB(t, db)
+
+	balPubkey := []byte{1, 2, 3}
+
+	proposalHistory, err := db.ProposalHistory(balPubkey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(proposalHistory, &slashpb.ProposalHistory{}) {
+		t.Fatalf("Expected proposal history to be empty, received: %v", proposalHistory)
+	}
+}
+
+func TestSaveProposalHistory_OK(t *testing.T) {
+	db := SetupDB(t, [][48]byte{})
+	defer TeardownDB(t, db)
+	tests := []struct {
+		pubkey  []byte
+		epoch   uint64
+		history *slashpb.ProposalHistory
+	}{
+		{
+			pubkey: []byte{0},
+			epoch:  uint64(1),
+			history: &slashpb.ProposalHistory{
+				EpochBits:          bitfield.Bitlist{0x02, 0x02},
+				LatestEpochWritten: 1,
+			},
+		},
+		{
+			pubkey: []byte{1},
+			epoch:  uint64(1),
+			history: &slashpb.ProposalHistory{
+				EpochBits:          bitfield.Bitlist{0x02, 0x02},
+				LatestEpochWritten: 1,
+			},
+		},
+		{
+			pubkey: []byte{3},
+			epoch:  uint64(2),
+			history: &slashpb.ProposalHistory{
+				EpochBits:          bitfield.Bitlist{0x04, 0x04},
+				LatestEpochWritten: 2,
+			},
+		},
+	}
+
+	for _, tt := range tests {
```

Comment:
General comment on testing: if you name your tests and run them with `t.Run()` they will run individually, so you can do things like exclude them, run them in parallel, run them individually, _etc_..  See `TestSecretKeyFromBytes()` in `shared/bls/bls_test.go` for an example.

Reply:
Good idea, thank you